Firestore/core/src/firebase/firestore/immutable/keys_view.h

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_KEYS_VIEW_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_KEYS_VIEW_H_
+
+#include "Firestore/core/src/firebase/firestore/util/iterator_adaptors.h"
+#include "Firestore/core/src/firebase/firestore/util/range.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+namespace impl {
+
+template <typename Iterator>
+util::range<util::iterator_first<Iterator>> MakeKeysRange(Iterator begin,
```

Comment:
Optional: this type is repeated quite a bit, perhaps define an alias?
```cpp
template <typename I>
using KeysRange = util::range<util::iterator_first<I>>;

template <typename Iterator>
KeysRange<Iterator> MakeKeysRange(...);
template <typename Range>
auto KeysView(...) -> KeysRange<decltype(std::begin(range))>;
// or perhaps
template <typename Range>
KeysRange<typename Range::iterator> KeysView(...);
```

Reply:
Did the first. Good idea.